provider/reprovider_test.go

```diff
@@ -221,3 +223,93 @@ func TestOfflineRecordsThenOnlineRepublish(t *testing.T) {
 		t.Fatalf("keys are not equal expected %v, got %v", someHash, prov.keys[0])
 	}
 }
+
+func newMockKeyChanFunc(cids []cid.Cid) KeyChanFunc {
+	return func(ctx context.Context) (<-chan cid.Cid, error) {
+		outCh := make(chan cid.Cid)
+
+		go func() {
+			defer close(outCh)
+			for _, c := range cids {
+				select {
+				case <-ctx.Done():
+					return
+				case outCh <- c:
+				}
+			}
+		}()
+
+		return outCh, nil
+	}
+}
+
+func makeCIDs(n int) []cid.Cid {
+	cids := make([]cid.Cid, n)
+	for i := 0; i < n; i++ {
+		buf := make([]byte, 63)
+		_, err := rand.Read(buf)
+		if err != nil {
+			panic(err)
+		}
+		data, err := mh.Encode(buf, mh.SHA2_256)
+		if err != nil {
+			panic(err)
+		}
+		cids[i] = cid.NewCidV1(0, data)
+	}
+
+	return cids
+}
+
+func TestNewPrioritizedProvider(t *testing.T) {
+	cids := makeCIDs(6)
+
+	testCases := []struct {
+		name     string
+		priority []cid.Cid
+		all      []cid.Cid
+		expected []cid.Cid
+	}{
+		{
+			name:     "basic test",
+			priority: cids[:3],
+			all:      cids[3:],
+			expected: cids,
+		},
+		{
+			name:     "basic test inverted",
+			priority: cids[3:],
+			all:      cids[:3],
+			expected: append(cids[3:], cids[:3]...),
+		},
+		{
+			name:     "no repeated",
+			priority: cids[3:],
+			all:      cids[3:],
+			expected: cids[3:],
+		},
+		{
+			name:     "no repeated intercalated",
```

Comment:
"intercalated" doesn't seem like the right word, but what's the test here "no repeated if duplicates in the prioritized channel"?